swarm/storage/netstore.go

```diff
@@ -18,318 +18,280 @@ package storage
 	return &NetStore{
-		Store:             store,
-		fetchers:          fetchers,
-		NewNetFetcherFunc: nnf,
-		closeC:            make(chan struct{}),
-	}, nil
+		Store:   store,
+		localID: localID,
+	}
 }
 
 // Put stores a chunk in localstore, and delivers to all requestor peers using the fetcher stored in
 // the fetchers cache
 func (n *NetStore) Put(ctx context.Context, mode chunk.ModePut, ch Chunk) (bool, error) {
-	n.mu.Lock()
-	defer n.mu.Unlock()
+	n.putMu.Lock()
+	defer n.putMu.Unlock()
+
+	log.Trace("netstore.put", "ref", ch.Address().String(), "mode", mode)
 
-	// put to the chunk to the store, there should be no error
+	// put the chunk to the localstore, there should be no error
 	exists, err := n.Store.Put(ctx, mode, ch)
 	if err != nil {
 		return exists, err
 	}
 
-	// if chunk is now put in the store, check if there was an active fetcher and call deliver on it
-	// (this delivers the chunk to requestors via the fetcher)
-	log.Trace("n.getFetcher", "ref", ch.Address())
-	if f := n.getFetcher(ch.Address()); f != nil {
-		log.Trace("n.getFetcher deliver", "ref", ch.Address())
-		f.deliver(ctx, ch)
-	}
-	return exists, nil
-}
+	// notify RemoteGet about a chunk being stored
+	fi, ok := n.fetchers.Load(ch.Address().String())
+	if ok {
+		// we need SafeClose, because it is possible for a chunk to both be
+		// delivered through syncing and through a retrieve request
+		fii := fi.(*FetcherItem)
+		fii.SafeClose()
+		log.Trace("netstore.put chunk delivered and stored", "ref", ch.Address().String())
 
-// Get retrieves the chunk from the NetStore DPA synchronously.
-// It calls NetStore.get, and if the chunk is not in local Storage
-// it calls fetch with the request, which blocks until the chunk
-// arrived or context is done
-func (n *NetStore) Get(rctx context.Context, mode chunk.ModeGet, ref Address) (Chunk, error) {
-	chunk, fetch, err := n.get(rctx, mode, ref)
-	if err != nil {
-		return nil, err
-	}
-	if chunk != nil {
-		// this is not measuring how long it takes to get the chunk for the localstore, but
-		// rather just adding a span for clarity when inspecting traces in Jaeger, in order
-		// to make it easier to reason which is the node that actually delivered a chunk.
-		_, sp := spancontext.StartSpan(
-			rctx,
-			"localstore.get")
-		defer sp.Finish()
-
-		return chunk, nil
-	}
-	return fetch(rctx)
-}
+		metrics.GetOrRegisterResettingTimer(fmt.Sprintf("netstore.fetcher.lifetime.%s", fii.CreatedBy), nil).UpdateSince(fii.CreatedAt)
 
-// FetchFunc returns nil if the store contains the given address. Otherwise it returns a wait function,
-// which returns after the chunk is available or the context is done
-func (n *NetStore) FetchFunc(ctx context.Context, ref Address) func(context.Context) error {
-	chunk, fetch, _ := n.get(ctx, chunk.ModeGetRequest, ref)
-	if chunk != nil {
-		return nil
-	}
-	return func(ctx context.Context) error {
-		_, err := fetch(ctx)
-		return err
+		if time.Since(fii.CreatedAt) > 5*time.Second {
```

Comment:
pull timeout to `timeouts`?

Reply:
@justelad this is not a timeout, just a helper snippet to log a reference for a chunk, that took way to long to be delivered, so that we inspect in code.

We can do the same through Jaeger and the tracing framework as well, but the reality is that the tracing setup is not scaling as good as our logs, hence the check here as well. I'd rather we just keep this like this for now, as it really is just debug code, and we can remove it when it is no longer needed.

I am adding a log message to explain this now.